File: android-companion/src/core/seed-card.ts
import { z } from "zod";
import type { CompanionState } from "./store";
import type { Message, SeedAnswerReceipt, SeedCardResult } from "./types";

// The native bundle stays independent of the server build. These constants
// mirror the versioned contract documented in docs/seed-answers.md.
export const SEED_CARD_PURPOSE = "onboarding-v1";
export const SEED_CARD_TITLE = "What do you mostly want help with?";
export const SEED_CARD_SUBTITLE = "Pick whatever's closest; we can always expand from there.";
export const SEED_CARD_OPTIONS = ["Work & projects", "Writing & research", "Life admin", "A bit of everything"];
export const seedAnswerTextSchema = z.string().max(4000).refine((answer) => answer.trim().length > 0);
export const seedAnswerReceiptSchema = z.object({
  messageId: z.string().min(1), attempt: z.number().int().nonnegative().max(Number.MAX_SAFE_INTEGER),
  status: z.enum(["recorded", "starting", "started", "not-started", "uncertain"]), error: z.string().optional(),
}).strict().refine((receipt) => receipt.status === "recorded" ? receipt.attempt === 0 : receipt.attempt > 0);

// Check original wire fields BEFORE the ordinary lossy decoder removes them.
export const seedWireMetadataSchema = z.object({
  role: z.literal("bot"), kind: z.literal("options"),
  parentId: z.string().nullable().optional(), from: z.null().optional(),
  card: z.object({
    title: z.string(), subtitle: z.string().optional(), options: z.array(z.string()),
    purpose: z.string().optional(), seedAnswer: seedAnswerReceiptSchema.optional(),
    answered: z.string().nullable().optional(), dismissed: z.boolean().optional(),
    requestId: z.string().optional(), tool: z.string().optional(), held: z.string().optional(), allowKey: z.string().optional(),
  }),
});

const answerCardSchema = z.object({
  title: z.literal(SEED_CARD_TITLE), subtitle: z.literal(SEED_CARD_SUBTITLE),
  options: z.array(z.string()).refine((options) => sameOptions(options)),
  purpose: z.literal(SEED_CARD_PURPOSE).optional(),
  seedAnswer: seedAnswerReceiptSchema.optional(), answered: z.string().optional(), dismissed: z.literal(false).optional(),
  requestId: z.never().optional(), tool: z.never().optional(), held: z.never().optional(), allowKey: z.never().optional(),
});
const answerMessageSchema = z.object({
  id: z.string().min(1), role: z.literal("bot"), kind: z.literal("options"), at: z.number().finite(),
  parentId: z.string().nullable().optional(), from: z.null().optional(), card: answerCardSchema,
});
const userMessageSchema = z.object({
  id: z.string().min(1), role: z.literal("user"), kind: z.literal("text"), at: z.number().finite(),
  parentId: z.string().nullable().optional(), from: z.null().optional(), text: seedAnswerTextSchema,
});
export const seedCardResultSchema = z.object({
  ok: z.literal(true),
  outcome: z.enum(["recorded", "already-recorded", "starting", "already-requested"]).optional(),
  cardMessage: answerMessageSchema, userMessage: userMessageSchema.nullable(),
}).refine((result) => {
  const card = result.cardMessage.card;
  if (!card.seedAnswer) return !result.outcome && !result.userMessage && card.answered === undefined;
  return !!result.userMessage && card.seedAnswer.messageId !== result.cardMessage.id
    && card.seedAnswer.messageId === result.userMessage.id && card.answered === result.userMessage.text;
});

function sameOptions(options: readonly string[]): boolean {
  return options.length === SEED_CARD_OPTIONS.length && options.every((option, index) => option === SEED_CARD_OPTIONS[index]);
}
function canonical(message: Message): boolean {
  const card = message.card;
  return !message.seedContextInvalid && message.role === "bot" && message.kind === "options" && !!card && !card.seedInvalid
    && message.from == null && card.requestId === undefined && card.tool === undefined && card.held === undefined && card.allowKey === undefined
    && card.title === SEED_CARD_TITLE && card.subtitle === SEED_CARD_SUBTITLE && sameOptions(card.options);
}
export function isRecognizedSeedCard(messages: readonly Message[], message: Message): boolean {
  const card = message.card;
  if (!card || !canonical(message) || card.dismissed || (card.answered != null && !card.seedAnswer)) return false;
  if (card.seedAnswer && (!seedAnswerReceiptSchema.safeParse(card.seedAnswer).success || !seedAnswerTextSchema.safeParse(card.answered).success)) return false;
  if (card.purpose === SEED_CARD_PURPOSE) return true;
  if (card.purpose !== undefined || messages[1]?.id !== message.id) return false;
  const greeting = messages[0];
  return !!greeting && !greeting.seedContextInvalid && greeting.role === "bot" && greeting.kind === "text" && greeting.from == null
    && greeting.parentId == null && message.parentId === greeting.id
    && /^Hey — I'm .+\. Nice to meet you\.$/u.test(greeting.text ?? "");
}

export function seedCardSignature(message: Message): string {
  const card = message.card;
  return JSON.stringify([message.id, message.parentId ?? null, card?.title, card?.subtitle, card?.options]);
}

/** Never use visibleTranscript's missing-leaf fallback to authorize a write. */
function activePath(state: CompanionState, threadId: string): Message[] {
  const all = state.messages[threadId] ?? [];
  const leaf = state.leaves[threadId] ?? all.at(-1)?.id;
  if (!leaf) return [];
  const byId = new Map(all.map((message) => [message.id, message]));
  const seen = new Set<string>();
  const result: Message[] = [];
  let message = byId.get(leaf);
  while (message) {
    if (seen.has(message.id)) return [];
    seen.add(message.id);
    result.push(message);
    message = message.parentId ? byId.get(message.parentId) : undefined;
  }
  return result.reverse();
}
export function seedCardOnActiveBranch(state: CompanionState, botId: string, threadId: string, cardId: string): Message | null {
  const bot = state.bots[botId];
  if (!bot || bot.hidden || bot.threadId !== threadId) return null;
  const all = state.messages[threadId] ?? [];
  const message = activePath(state, threadId).find((candidate) => candidate.id === cardId);
  return message && isRecognizedSeedCard(all, message) ? message : null;
}
export function seedCardWriteBlocker(state: CompanionState, botId: string, threadId: string, cardId: string): string | null {
  const message = seedCardOnActiveBranch(state, botId, threadId, cardId);
  if (!message?.card) return "Open the current bot conversation to answer this question.";
  if (state.bots[botId].busy) return "This bot is working. Wait for it to finish before starting this saved task.";
  const path = activePath(state, threadId);
  const anchor = path.findIndex((candidate) => candidate.id === (message.card?.seedAnswer?.messageId ?? cardId));
  if (anchor < 0) return "The saved answer is not on this branch. Check status to confirm its record.";
  if (message.card.seedAnswer) {
    const linked = path[anchor];
    if (linked.role !== "user" || linked.kind !== "text" || linked.text !== message.card.answered) return "The saved answer could not be confirmed. Check status before starting it.";
  }
  if (path.slice(anchor + 1).some((candidate) => candidate.role === "user")) return "This conversation contains newer work. This earlier question can no longer start a task.";
  return null;
}

export function matchesSeedCardResult(card: Message, result: SeedCardResult): boolean {
  const parsed = seedCardResultSchema.safeParse(result);
  return parsed.success && card.id === result.cardMessage.id && seedCardSignature(card) === seedCardSignature(result.cardMessage);
}
function receiptProgress(receipt: SeedAnswerReceipt): number {
  return receipt.status === "recorded" ? 0 : receipt.status === "starting" ? 1 : 2;
}
/** API may finish before an older SSE patch arrives. Neither source may undo it. */
export function preserveSeedReceipt(current: Message, incoming: Message): Message {
  const before = current.card?.seedAnswer;
  if (!before || !canonical(current)) return incoming;
  const after = incoming.card?.seedAnswer;
  if (!canonical(incoming) || seedCardSignature(current) !== seedCardSignature(incoming) || !after
    || !seedAnswerReceiptSchema.safeParse(after).success || after.messageId !== before.messageId
    || incoming.card?.answered !== current.card?.answered || after.attempt < before.attempt
    || (after.attempt === before.attempt && (receiptProgress(after) < receiptProgress(before)
      || receiptProgress(before) === 2 && after.status !== before.status))) return current;
  return incoming;
}

/** Only a new direct-child echo advances the leaf; newer branches and streams survive. */
export function mergeSeedCardResult(state: CompanionState, botId: string, threadId: string, cardId: string, result: SeedCardResult): CompanionState {
  const current = seedCardOnActiveBranch(state, botId, threadId, cardId);
  if (!current?.card || !matchesSeedCardResult(current, result) || preserveSeedReceipt(current, result.cardMessage) === current) return state;
  const all = state.messages[threadId] ?? [];
  const echo = result.userMessage;
  const existing = echo ? all.find((message) => message.id === echo.id) : undefined;
  if (echo && existing && (existing.role !== "user" || existing.kind !== "text" || existing.text !== echo.text || existing.parentId !== echo.parentId)) return state;
  const card = result.cardMessage.card!;
  const next = all.map((message) => message.id === cardId ? {
    ...message, card: { ...message.card!, purpose: card.purpose ?? message.card?.purpose, answered: card.answered, seedAnswer: card.seedAnswer },
  } : message);
  if (echo && !existing) {
    const branch = activePath(state, threadId);
    const parentIndex = branch.findIndex((message) => message.id === echo.parentId);
    const cardIndex = branch.findIndex((message) => message.id === cardId);
    if (parentIndex < cardIndex) return state;
    next.push(echo);
  }
  const leaf = state.leaves[threadId] ?? all.at(-1)?.id;
  const advances = echo && !existing && echo.parentId === leaf;
  return { ...state, messages: { ...state.messages, [threadId]: next },
    leaves: advances ? { ...state.leaves, [threadId]: echo.id } : state.leaves };
}

File: ios/Sources/CompanionCore/SeedCard.swift
import Foundation

public enum SeedAnswerStatus: String, Codable, Hashable, Sendable {
    case recorded, starting, started, uncertain
    case notStarted = "not-started"
}

public struct SeedAnswerReceipt: Codable, Hashable, Sendable {
    public var messageId: String
    public var attempt: Int
    public var status: SeedAnswerStatus
    public var error: String?

    public init(messageId: String, attempt: Int, status: SeedAnswerStatus, error: String? = nil) {
        self.messageId = messageId; self.attempt = attempt; self.status = status; self.error = error
    }
    public var isValid: Bool {
        !messageId.isEmpty && (0...SeedCardContract.maximumAttempt).contains(attempt)
            && (status == .recorded ? attempt == 0 : attempt > 0)
    }
    private enum CodingKeys: String, CodingKey { case messageId, attempt, status, error }
    public init(from decoder: Decoder) throws {
        let raw = try decoder.container(keyedBy: SeedJSONKey.self)
        guard Set(raw.allKeys.map(\.stringValue)).isSubset(of: ["messageId", "attempt", "status", "error"]) else {
            throw SeedCardContract.invalid(decoder, "Unknown welcome receipt field")
        }
        let c = try decoder.container(keyedBy: CodingKeys.self)
        messageId = try c.decode(String.self, forKey: .messageId)
        attempt = try c.decode(Int.self, forKey: .attempt)
        status = try c.decode(SeedAnswerStatus.self, forKey: .status)
        error = c.contains(.error) ? try c.decode(String.self, forKey: .error) : nil
        guard isValid else { throw SeedCardContract.invalid(decoder, "Invalid welcome receipt") }
    }
}

private struct SeedJSONKey: CodingKey {
    var stringValue: String
    var intValue: Int? { nil }
    init?(stringValue: String) { self.stringValue = stringValue }
    init?(intValue: Int) { return nil }
}

public struct SeedCardResult: Decodable, Sendable {
    public var ok: Bool
    public var outcome: String?
    public var cardMessage: Message
    public var userMessage: Message?

    public init(ok: Bool = true, outcome: String? = nil, cardMessage: Message, userMessage: Message?) {
        self.ok = ok; self.outcome = outcome; self.cardMessage = cardMessage; self.userMessage = userMessage
    }
    private enum CodingKeys: String, CodingKey { case ok, outcome, cardMessage, userMessage }
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        ok = try c.decode(Bool.self, forKey: .ok)
        outcome = c.contains(.outcome) ? try c.decode(String.self, forKey: .outcome) : nil
        cardMessage = try c.decode(Message.self, forKey: .cardMessage)
        // Missing userMessage differs from an explicitly unanswered status.
        guard c.contains(.userMessage) else { throw SeedCardContract.invalid(decoder, "Missing welcome user message") }
        userMessage = try c.decodeIfPresent(Message.self, forKey: .userMessage)
        let rawMessage = try c.nestedContainer(keyedBy: SeedJSONKey.self, forKey: .cardMessage)
        let rawCard = try rawMessage.nestedContainer(keyedBy: SeedJSONKey.self, forKey: SeedJSONKey(stringValue: "card")!)
        let answered = SeedJSONKey(stringValue: "answered")!
        if rawCard.contains(answered) { _ = try rawCard.decode(String.self, forKey: answered) }
        guard isValid else { throw SeedCardContract.invalid(decoder, "Invalid welcome answer response") }
    }

    public var isValid: Bool {
        guard ok, outcome == nil || ["recorded", "already-recorded", "starting", "already-requested"].contains(outcome!),
              SeedCardContract.canonical(cardMessage), let card = cardMessage.card,
              card.purpose == nil || card.purpose == SeedCardContract.purpose, card.dismissed != true else { return false }
        guard let receipt = card.seedAnswer else { return outcome == nil && userMessage == nil && card.answered == nil }
        guard receipt.isValid, let echo = userMessage, !echo.seedContextInvalid, echo.role == .user,
              echo.kind == .text, echo.from == nil, echo.at.isFinite, !echo.id.isEmpty,
              echo.id != cardMessage.id, echo.id == receipt.messageId,
              let text = echo.text, SeedCardContract.isValidAnswer(text), let answer = card.answered,
              SeedCardContract.exactText(text, answer) else { return false }
        return true
    }
}

public enum SeedMergeOutcome: Equatable, Sendable { case accepted, rejected }

public enum SeedCardContract {
    public static let purpose = "onboarding-v1"
    public static let title = "What do you mostly want help with?"
    public static let subtitle = "Pick whatever's closest; we can always expand from there."
    public static let options = ["Work & projects", "Writing & research", "Life admin", "A bit of everything"]
    // The server's attempt counter is a JS number, so the valid domain ends
    // at 2^53−1. The watchOS simulator is arm64_32, where Int.max is smaller
    // than that — and smaller than any attempt count that can ever exist —
    // so the clamp is semantically identical: "no real attempt reaches this".
    public static let maximumAttempt: Int = Int(min(Double(Int.max), 9_007_199_254_740_991))

    public static func exactText(_ lhs: String, _ rhs: String) -> Bool {
        lhs.utf16.elementsEqual(rhs.utf16)
    }
    public static func isValidAnswer(_ text: String) -> Bool {
        // ECMAScript trim's WhiteSpace + LineTerminator set, including BOM.
        text.utf16.count <= 4000 && text.unicodeScalars.contains { scalar in
            switch scalar.value {
            case 0x0009...0x000D, 0x0020, 0x00A0, 0x1680, 0x2000...0x200A, 0x2028, 0x2029, 0x202F, 0x205F, 0x3000, 0xFEFF: return false
            default: return true
            }
        }
    }
    static func invalid(_ decoder: Decoder, _ description: String) -> DecodingError {
        .dataCorrupted(.init(codingPath: decoder.codingPath, debugDescription: description))
    }
    static func canonical(_ message: Message) -> Bool {
        guard !message.seedContextInvalid, !message.id.isEmpty, message.at.isFinite,
              message.role == .bot, message.kind == .options, message.from == nil,
              let card = message.card, !card.seedInvalid, card.requestId == nil, card.tool == nil,
              card.held == nil, card.allowKey == nil,
              exactText(card.title, title), exactText(card.subtitle, subtitle), card.options.count == options.count else { return false }
        return zip(card.options, options).allSatisfy { exactText($0, $1) }
    }
    public static func isRecognized(_ message: Message, in messages: [Message]) -> Bool {
        guard canonical(message), let card = message.card, card.dismissed != true,
              card.answered == nil || card.seedAnswer != nil else { return false }
        if let receipt = card.seedAnswer {
            guard receipt.isValid, let answer = card.answered, isValidAnswer(answer) else { return false }
        }
        if card.purpose == purpose { return true }
        guard card.purpose == nil, messages.count > 1, messages[1].id == message.id else { return false }
        let greeting = messages[0]
        guard !greeting.seedContextInvalid, greeting.role == .bot, greeting.kind == .text,
              greeting.from == nil, greeting.parentId == nil, message.parentId == greeting.id,
              let text = greeting.text else { return false }
        let prefix = "Hey — I'm ", suffix = ". Nice to meet you."
        guard text.hasPrefix(prefix), text.hasSuffix(suffix) else { return false }
        let name = text.dropFirst(prefix.count).dropLast(suffix.count)
        return !name.isEmpty && !name.unicodeScalars.contains { [10, 13, 0x2028, 0x2029].contains(Int($0.value)) }
    }
    public static func signature(_ message: Message) -> String {
        // JSON bytes encode canonically distinct Unicode without Swift's String equality collapsing it.
        let fields: [String?] = [message.id, message.parentId, message.card?.title, message.card?.subtitle]
            + (message.card?.options.map(Optional.some) ?? [])
        return ((try? JSONEncoder().encode(fields)) ?? Data()).base64EncodedString()
    }
    public static func matches(_ current: Message, _ result: SeedCardResult) -> Bool {
        result.isValid && current.id == result.cardMessage.id && signature(current) == signature(result.cardMessage)
    }
    private static func progress(_ receipt: SeedAnswerReceipt) -> Int {
        receipt.status == .recorded ? 0 : receipt.status == .starting ? 1 : 2
    }
    static func preserveReceipt(_ current: Message, _ incoming: Message) -> Message {
        guard canonical(current), let before = current.card?.seedAnswer, before.isValid else { return incoming }
        guard canonical(incoming), signature(current) == signature(incoming), let after = incoming.card?.seedAnswer,
              after.isValid, after.messageId == before.messageId,
              let oldAnswer = current.card?.answered, let newAnswer = incoming.card?.answered, exactText(oldAnswer, newAnswer),
              after.attempt >= before.attempt else { return current }
        if after.attempt == before.attempt && (progress(after) < progress(before)
            || progress(before) == 2 && after.status != before.status) { return current }
        return incoming
    }
}

extension CompanionState {
    /// Unlike display fallbacks, an unknown leaf or cycle never authorizes an action.
    func seedActivePath(_ threadId: String) -> [Message] {
        let all = transcript(forThread: threadId)
        guard let leaf = bot(forThread: threadId)?.activeLeafId ?? all.last?.id else { return [] }
        let byId = Dictionary(all.map { ($0.id, $0) }, uniquingKeysWith: { _, new in new })
        guard var current = byId[leaf] else { return [] }
        var path: [Message] = [], seen = Set<String>()
        while true {
            guard seen.insert(current.id).inserted else { return [] }
            path.append(current)
            guard let parentId = current.parentId, let parent = byId[parentId] else { break }
            current = parent
        }
        return path.reversed()
    }
    public func seedCard(botId: String, threadId: String, cardId: String) -> Message? {
        guard let bot = bot(botId), bot.hidden != true, bot.threadId == threadId,
              room(forThread: threadId) == nil,
              let card = seedActivePath(threadId).first(where: { $0.id == cardId }),
              SeedCardContract.isRecognized(card, in: transcript(forThread: threadId)) else { return nil }
        return card
    }
    public func seedWriteBlocker(botId: String, threadId: String, cardId: String) -> String? {
        guard let message = seedCard(botId: botId, threadId: threadId, cardId: cardId), let card = message.card else {
            return "Open the current bot conversation to answer this question."
        }
        if bot(botId)?.busy == true { return "This bot is working. Wait for it to finish before starting this saved task." }
        let path = seedActivePath(threadId)
        guard let anchor = path.firstIndex(where: { $0.id == (card.seedAnswer?.messageId ?? cardId) }) else {
            return "The saved answer is not on this branch. Check status to confirm its record."
        }
        if card.seedAnswer != nil {
            let echo = path[anchor]
            guard echo.role == .user, echo.kind == .text, let text = echo.text, let answer = card.answered,
                  SeedCardContract.exactText(text, answer) else { return "The saved answer could not be confirmed. Check status before starting it." }
        }
        if path.dropFirst(anchor + 1).contains(where: { $0.role == .user }) {
            return "This conversation contains newer work. This earlier question can no longer start a task."
        }
        return nil
    }
    public func unresolvedSeedAnswerUser(threadId: String, messageId: String) -> Bool {
        transcript(forThread: threadId).contains { message in
            guard SeedCardContract.isRecognized(message, in: transcript(forThread: threadId)),
                  let receipt = message.card?.seedAnswer else { return false }
            return receipt.messageId == messageId && receipt.status != .started
        }
    }
    public mutating func mergeSeedAnswer(botId: String, threadId: String, cardId: String, result: SeedCardResult) -> SeedMergeOutcome {
        guard let current = seedCard(botId: botId, threadId: threadId, cardId: cardId),
              SeedCardContract.matches(current, result) else { return .rejected }
        var all = transcript(forThread: threadId)
        let path = seedActivePath(threadId)
        guard let cardIndex = path.firstIndex(where: { $0.id == cardId }) else { return .rejected }
        let echo = result.userMessage
        let existing = echo.flatMap { incoming in all.first { $0.id == incoming.id } }
        if let echo {
            guard let parent = echo.parentId, let parentIndex = path.firstIndex(where: { $0.id == parent }), parentIndex >= cardIndex else { return .rejected }
            if existing == nil && echo.parentId != path.last?.id { return .rejected }
            if let existing {
                guard !existing.seedContextInvalid, existing.role == .user, existing.kind == .text, existing.from == nil,
                      let text = existing.text, let incomingText = echo.text, SeedCardContract.exactText(text, incomingText),
                      existing.parentId == echo.parentId, let index = path.firstIndex(where: { $0.id == echo.id }), index > cardIndex else { return .rejected }
            }
        }
        let selected = SeedCardContract.preserveReceipt(current, result.cardMessage)
        // A newer local receipt may win, but only for the same exact saved answer and echo.
        if let before = current.card?.seedAnswer {
            guard let after = result.cardMessage.card?.seedAnswer, before.messageId == after.messageId,
                  let text = current.card?.answered, let incomingText = result.cardMessage.card?.answered,
                  SeedCardContract.exactText(text, incomingText) else { return .rejected }
            let beforeTerminal = before.status != .recorded && before.status != .starting
            let afterTerminal = after.status != .recorded && after.status != .starting
            if before.attempt == after.attempt && beforeTerminal && afterTerminal && before.status != after.status { return .rejected }
        }
        guard let index = all.firstIndex(where: { $0.id == cardId }), let card = selected.card else { return .rejected }
        all[index].card?.purpose = card.purpose ?? current.card?.purpose
        all[index].card?.answered = card.answered
        all[index].card?.seedAnswer = card.seedAnswer
        if let echo, existing == nil {
            all.append(echo)
            if echo.parentId == path.last?.id, let botIndex = bots.firstIndex(where: { $0.id == botId }) {
                bots[botIndex].activeLeafId = echo.id
            }
        }
        messages[threadId] = all
        return .accepted
    }
}
